### PR TITLE
add feature-based and gradient-based similarity methods

### PR DESCRIPTION
**Pull request by LIshuaike**

### CODE REVIEW

File: trustai/interpretation/example_level/README.md [hunk 1]
```diff
@@ -43,4 +43,89 @@ text: 价格确实比较高，而且还没有早餐提供。 携程拿到的价
...
+predict_labels, most_similar_examples, most_dissimilar_examples = feature_sim_model(test_dataloader, sample_num=3, sim_fn='cos')
+```
+
+结果示例：
+```txt
+text: 本来不想评价了，但为了携程的携粉们，还是说一下，这称不上是九点，细说就真没必要了，就一个字：差    predict label: 0
+most similar examples
+text: 我选分期付款，在上海扣的款，在上海开的发票，东西却从北京用快递发出，真是舍近求远，害的我等了一星期才收到货，真是脑残！    gold label: 0
+text: 看到评价那么高，就买了，但女儿不喜欢，我看了一下，也不喜欢，不知所云，汽车什么的，都是过时的，或生活中没有的。不明白，为什么有这么高的评价。    gold label: 0
+text: 看到评价那么高，就买了，但女儿不喜欢，我看了一下，也不喜欢，不知所云，汽车什么的，都是过时的，或生活中没有的。不明白，为什么有这么高的评价。    gold label: 0
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/README.md]: case 挑点没有重复的吧，不然用户会很奇怪为什么挑出来相似的case。
特别是选的case尽量好一点。 特别是如果表示点选出来的不好，而gc和fc又一模一样的话，调用示例和输出case可以只选择fc或者gc。

File: trustai/interpretation/example_level/method/representer_point.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 from scipy.stats.stats import pearsonr
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/representer_point.py]: 这个搞成lazy import吧， 不然用户还需要强制安装scipy这个库，增加了使用难度。  
可以参考在/trustai/interpretation/token_level/method/norm_lime.py中的实现，但是记得错误警告只能提示一次。也可以使用warnings库。

File: trustai/interpretation/example_level/method/representer_point.py [hunk 3]
```diff
@@ -57,20 +58,24 @@ class RepresenterPointBase(nn.Layer):
...
+    ):
         """
         Initialization
         """
         super().__init__()
         weight, params = self._get_params(paddle_model, classifier_layer_name)
-        self.softmax_classifier = SoftmaxClassifier(weight.shape[0] + 1, weight.shape[1], params)
+        self.softmax_classifier = SoftmaxClassifier(
+            weight.shape[0] + 1, weight.shape[1], params
+        )
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/representer_point.py]: 代码风格检查的长度要设置为120，80或者太短了，读起来也不友好。

File: trustai/interpretation/example_level/method/feature_similarity.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        ExampleBaseInterpreter.__init__(
+            self, paddle_model, device, predict_fn, classifier_layer_name
+        )
+        self.paddle_model = paddle_model
+        self.classifier_layer_name = classifier_layer_name
+        self.train_feature, _ = self.extract_featue(paddle_model, train_dataloader)
+
+    def interpret(self, data, sample_num=3, sim_fn="dot"):
+        """
+        Select most similar examples for a given data using the `sim_fn` metric.
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/feature_similarity.py]: 还返回了不相似

File: trustai/interpretation/example_level/method/feature_similarity.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        Args:
+            data(iterable): Dataloader to interpret.
+            sample_sum(int: default=3): the number of positive examples and negtive examples selected for each instance.
+            sim_fn(str: default=dot): the similarity metric to select examples.
+        """
+        examples = []
+        val_feature, preds = self.extract_featue(self.paddle_model, data)
+        if sim_fn == "dot":
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/feature_similarity.py:54]: 这里之前不是讨论过每个相似度的方法都抽象出来，这样代码的可读性也会好很多么 你这么写重复代码太多了
另外如果用户输入不存在的相似度函数名称，你需要警告用户目前只支持 dot cos euc
```python
if sim_fn == "dot":
     similarity_function = funxxx1
elif sim_fn == "cos":
     similarity_function = funxxx2
elif sim_fn == "euc:
     similarity_function = funxxx3
else:
      warning(xxx)
for index, target_class in enumerate(preds):
       tmp = similarity_function(xxxx)
       example_index = self._get_similarity_index(tmp, sample_num=sample_num)
       examples.append(example_index)
```
或者，你还可以设置一个function_dict，根据用户传入的sim_fn取值即可。

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        ExampleBaseInterpreter.__init__(
+            self, paddle_model, device, predict_fn, classifier_layer_name
+        )
+        self.paddle_model = paddle_model
+        self.classifier_layer_name = classifier_layer_name
+        self.criterion = (
+            criterion if criterion is not None else paddle.nn.loss.CrossEntropyLoss()
+        )
+        if cached_train_grad is not None:
+            self.train_feature = paddle.load(cached_train_grad)
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py]: 梯度的缓存为啥是train_feature

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            self.train_feature, *_ = self.get_grad(paddle_model, train_dataloader)
+            paddle.save(self.train_feature, "./cached_train_grad.tensor")
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py]: 1. 不要在用户目录下产出临时文件，缓存应该保存在tmp目录下。可以使用tempfile.gettempdir() 获取temp目录的位置。
2.cached_train_grad.tensor可以让用户自定义，防止不同模型冲突，另外cached_train_grad描述里面写的str类型。
3.考虑save失败的情况，保存失败不应该core掉。

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        Select most similar examples for a given data using the `sim_fn` metric.
+        Args:
+            data(iterable): Dataloader to interpret.
+            sample_sum(int: default=3): the number of positive examples and negtive examples selected for each instance.
+            sim_fn(str: default=dot): the similarity metric to select examples.
+        """
+        examples = []
+        val_feature, _, preds = self.get_grad(self.paddle_model, data)
+
+        if sim_fn == "dot":
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py:70]: 此处修改方法与feature相同。

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        get grad for data_loader.
+        """
+        paddle_model.eval()
+        print("Extracting gradient for given dataloader, it will take some time...")
+        features, probas, preds, grads = [], [], [], []
+
+        for step, batch in enumerate(data_loader, start=1):
+            *input, label = batch
+            _, prob, pred = self.predict_fn(input)
+            loss = self.criterion(prob, label)
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py]: 计算梯度应该用预测才对，为什么用的是golden label。

另外不需要计算loss，prob[predict_index].backward(retain_graph=False)即可。

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            probas.extend(prob)
+            preds.extend(pred)
+
+        return (
+            paddle.to_tensor(grads),
+            paddle.to_tensor(probas),
+            paddle.to_tensor(preds),
+        )
+
+    def _get_flat_param_grad(self):
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py:109]: 为什么使用的是classifier_layer的grad, 我理解不该是cls_emb的grad么

File: tests/interpretation/example_level/test_gradient_similarity.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
+# !/usr/bin/env python3
+import os
```
Comment (zhangyimi) [inline tests/interpretation/example_level/test_gradient_similarity.py:2]: 你单测写的太简单了，接口那么多参数能确保都没有问题么

**Review by zhangyimi — CHANGES_REQUESTED** [review]

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 12]
```diff
@@ -36,25 +38,25 @@ def __init__(
...
+        ExampleBaseInterpreter.__init__(self, paddle_model, device, predict_fn, classifier_layer_name)
         self.paddle_model = paddle_model
         self.classifier_layer_name = classifier_layer_name
-        self.criterion = (
-            criterion if criterion is not None else paddle.nn.loss.CrossEntropyLoss()
-        )
-        if cached_train_grad is not None:
-            self.train_feature = paddle.load(cached_train_grad)
+        self.criterion = (criterion if criterion is not None else paddle.nn.loss.CrossEntropyLoss())
+        if os.path.exists(cached_train_grad) and os.path.isfile(cached_train_grad):
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py]: cached_train_grad默认为None, 不然跑完一个模型，跑第二个模型的时候就出错了。

File: trustai/interpretation/example_level/method/feature_similarity.py [hunk 13]
```diff
@@ -86,6 +74,15 @@ def _get_similarity_index(self, scores, sample_num=3):
         dis_sim_index = index[-sample_num:].tolist()
         return sim_index, dis_sim_index
 
+    def _dot_similarity(self, inputs):
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/feature_similarity.py]: 这三个方法放到common的utils下吧， 没必要重复定义。
计算相似度本身也与这个类无关。

File: trustai/interpretation/example_level/method/representer_point.py [hunk 14]
```diff
@@ -148,23 +142,23 @@ def train(self, input_feature, input_probas):
         weight_matrix = weight_matrix / (-2.0 * self.lmbd * input_num)  # alpha
 
         if self.correlation:
-            best_w = paddle.matmul(
-                paddle.t(input_feature), weight_matrix
-            )  # alpha * f_i^T
+            try:
+                from scipy.stats.stats import pearsonr
+            except ImportError as e:
+                import sys
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/representer_point.py:148]: 日志打印一次即可，你这样每一次调用interpret函数都会打印一行日志。

**Review by zhangyimi — CHANGES_REQUESTED** [review]

File: trustai/interpretation/example_level/README.md [hunk 15]
```diff
@@ -45,80 +45,82 @@ text: 价格确实比较高，而且还没有早餐提供。 携程拿到的价
...
-most similar examples
+examples with positive influence
 text: Linux系统不太好用,平时习惯用Windows xp 系统,一下子用这个系统感觉很不习惯,建议开发或预装Windows xp系统.    gold label: 0
 text: 1、机器较沉 2、VISTA用起来不习惯，且占系统较多 3、音频插口、右前侧的二个USB口在用鼠标时感觉手靠得太近了    gold label: 0
 text: vista系统下也没有无线网卡驱动，用驱动精灵可解决。 机器稍有点重。 散热有待改进。    gold label: 0
-most dissimilar examples
-text: 分区时出现拒绝访问，怎么分都不行，请求帮助！ 送的1G内存条，怎么安装呢？应该不能自己擅自安装吧    gold label: 1
-text: 进BIOS，configuration 改成IDE接口就行！ 散热不好！太热了，建议价钱或者少出点钱购买其他款机器！    gold label: 1
-text: 没有配置驱动光盘。这点很不好。虽然说笔记本本身带刻录。可是也麻烦了客户啊。    gold label: 1
+examples with negative influence
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/README.md:80]: 可以加个说明，解释一下负例，第一条明显标注错误了。  后两条标注还冲突了。

File: trustai/interpretation/example_level/README.md [hunk 16]
```diff
@@ -45,80 +45,82 @@ text: 价格确实比较高，而且还没有早餐提供。 携程拿到的价
...
 ```txt
+test data
 text: 没有光驱,重装Windows需要外接光驱,对于电脑新手会很麻烦(没什么人会用Linux吧)    predict label: 0
-most similar examples
+examples with positive influence
 text: Linux系统不太好用,平时习惯用Windows xp 系统,一下子用这个系统感觉很不习惯,建议开发或预装Windows xp系统.    gold label: 0
 text: 1、机器较沉 2、VISTA用起来不习惯，且占系统较多 3、音频插口、右前侧的二个USB口在用鼠标时感觉手靠得太近了    gold label: 0
 text: vista系统下也没有无线网卡驱动，用驱动精灵可解决。 机器稍有点重。 散热有待改进。    gold label: 0
-most dissimilar examples
+examples with negative influence
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/README.md:123]: 在后面可以解释下基于fc方法选出来负例与gc有什么区别， fc倾向于选择字面不相似的数据作为负影响。

File: trustai/interpretation/example_level/method/gradient_similarity.py [hunk 17]
```diff
@@ -38,21 +38,22 @@ def __init__(
...
             criterion(paddle.nn.layer.loss: default=None): criterion to calculate model loss.
-            cached_train_grad(str: default="./cached_train_grad.tensor"): path of the cached train_dataloader gradient.
+            cached_train_grad(str: default=None): Path of the cached train_dataloader gradient. In the first training time, it will take some time to generate the train_grad
         """
         ExampleBaseInterpreter.__init__(self, paddle_model, device, predict_fn, classifier_layer_name)
         self.paddle_model = paddle_model
         self.classifier_layer_name = classifier_layer_name
         self.criterion = (criterion if criterion is not None else paddle.nn.loss.CrossEntropyLoss())
-        if os.path.exists(cached_train_grad) and os.path.isfile(cached_train_grad):
+        if cached_train_grad is not None and os.path.exists(cached_train_grad) and os.path.isfile(cached_train_grad):
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/method/gradient_similarity.py]: 调用isfile，就不用再调用exists了，当文件不存在时isfile会返回False。

**Review by zhangyimi — CHANGES_REQUESTED** [review]

File: trustai/interpretation/example_level/README.md [hunk 18]
```diff
@@ -127,7 +127,7 @@ text: 我从06年开始几乎每月都有一次出差，山西很多酒店都住
 ```
 
 
-基于特征的相似度方法召回了在特征意义上与测试数据最相似和最不相似的实例数据。
+基于特征的相似度方法召回了在特征意义上对测试数据有正影响和负影响的实例数据。召回的产生正影响的数据与GC方法一致，产生负影响的数据更倾向于和测试数据字面不相似的数据。
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/README.md]: 一致->相似

File: trustai/interpretation/example_level/README.md [hunk 19]
```diff
@@ -84,7 +84,7 @@ text: 买机器送的移动硬盘2.5寸250G的，没开封，想卖出，感兴
 ```
 
 
-基于梯度的相似度方法召回了在梯度意义上对测试数据有正影响和负影响的实例数据。
+基于梯度的相似度方法召回了在梯度意义上对测试数据有正影响和负影响的实例数据。召回的产生正影响的数据与训练数据语义上比较相似且标签一致，产生负影响的数据是标注错误的或者是存在冲突的。
```
Comment (zhangyimi) [inline trustai/interpretation/example_level/README.md]: “召回的产生正影响的数据与训练数据语义比较相似且标签一致” 这是病句吧
1. 召回的产生正影响 改为 召回的正影响数据 更通顺一点。下面特征部分也这么改。
2. 与训练数据比较相似，  召回的数据不就是训练数据么

产生负影响的数据是标注错误的或者是存在冲突的  这句话太绝对，我们只是给出的case有这种现象。
1.可以说：在基于梯度的方法中，召回的负影响数据通常可能存在如类别模糊、标注错误甚至矛盾的现象。

**Review by zhangyimi — APPROVED** [review]
